algebird-core/src/main/scala/com/twitter/algebird/HyperLogLog.scala

```diff
@@ -240,6 +240,65 @@ sealed abstract class HLL extends java.io.Serializable {
    * Set each item in the given buffer to the max of this and the buffer
    */
   def updateInto(buffer: Array[Byte]): Unit
+
+  /**
+   * Common accessor for finding the new rhoW value at j in downsized HLL
+   *
+   * [[http://research.neustar.biz/2012/09/12/set-operations-on-hlls-of-different-sizes/]]
+   *
+   * [[http://research.neustar.biz/2013/03/25/hyperloglog-engineering-choosing-the-right-bits/]]
+   *
+   * @param newJ        j for which the new rhoW is needed
+   * @param reducedBits Number of bits in the reduced HLL
+   * @param getRhoW     Function that returns the current rhoW at j
+   *
+   * @return  The new rhoW at j for an HLL with reducedBits
+   */
+  protected def getNewRhoWAt(newJ: Int, reducedBits: Int, getRhoW: Int => Byte): Byte = {
+    val reducedSize = 1 << reducedBits
+
+    // bit mask to set MSBs to 1. Makes rhoW exit fast
+    val bitMask = 0xFFFFFFFF << bits
+    // for j in HLL, we need to find the max of rhoW
+    // at current j, j + stepSize, ... j + (2^reducedBits - 1) * stepSize
+    val stepSize = 1 << reducedBits
+
+    // initialize offset and max
+    var offset = 0
+    var maxRhoW = 0: Byte
+    // iterate through to find the max over current rhoW
+    while (offset < size) {
+      val modifiedRhoW = {
+        val currentJ = newJ + offset
+        val currentRhoW = getRhoW(currentJ)
+        if (currentRhoW == 0)
+          // register not set, skip
+          currentRhoW
+        else if (currentJ < reducedSize)
+          // the MSBs are guaranteed to be all zero, so we can just add the difference to rhoW
+          (currentRhoW + (bits - reducedBits)).toByte
+        else {
+          // find the number of leading zeros
+          // we use bitMask to force rhoW to stop after going through (bits - reducedBits) bits
+          val buf = new Array[Byte](4)
```

Comment:
might as well allocate this outside the while loop so you can share across iterations for sure.